cli/src/commands/rebase.rs

```diff
@@ -348,92 +351,99 @@ fn rebase_descendants_transaction(
+    // descendants.
     tx.mut_repo().transform_descendants(
         settings,
-        vec![to_rebase_commit.id().clone()],
+        target_roots.iter().cloned().collect_vec(),
         |mut rewriter| {
             let old_commit = rewriter.old_commit();
             let old_commit_id = old_commit.id().clone();
 
-            // Replace references to `to_rebase_commit` with its parents.
-            rewriter.replace_parent(to_rebase_commit.id(), to_rebase_commit.parent_ids());
+            if target_roots.contains(&old_commit_id) {
+                rewriter.set_new_parents(new_parents.clone());
+            }
             if rewriter.parents_changed() {
                 let builder = rewriter.rebase(settings)?;
                 let commit = builder.write()?;
-                rebased_commit_ids.insert(old_commit_id, commit.id().clone());
+                new_rebased_commit_ids.insert(old_commit_id, commit.id().clone());
+            } else {
+                skipped_commits.push(rewriter.old_commit().clone());
             }
             Ok(())
         },
     )?;
 
-    let num_rebased_descendants = rebased_commit_ids.len();
-
-    // We now update `new_parents` to account for the rebase of all of
-    // `to_rebase_commit`'s descendants. Even if some of the original `new_parents`
-    // were descendants of `to_rebase_commit`, this will no longer be the case after
-    // the update.
-    let new_parents = new_parents
-        .iter()
-        .map(|new_parent| {
-            rebased_commit_ids
-                .get(new_parent.id())
-                .unwrap_or(new_parent.id())
-                .clone()
-        })
-        .collect();
-
-    let tx_description = format!("rebase commit {}", to_rebase_commit.id().hex());
-    // Finally, it's safe to rebase `to_rebase_commit`. We can skip rebasing if it
-    // is already a child of `new_parents`. Otherwise, at this point, it should
-    // no longer have any children; they have all been rebased and the originals
-    // have been abandoned.
-    let skipped_commit_rebase = if to_rebase_commit.parent_ids() == new_parents {
-        if let Some(mut formatter) = ui.status_formatter() {
-            write!(formatter, "Skipping rebase of commit ")?;
-            tx.write_commit_summary(formatter.as_mut(), &to_rebase_commit)?;
-            writeln!(formatter)?;
+    if let Some(mut fmt) = ui.status_formatter() {
+        if !skipped_commits.is_empty() {
+            if skipped_commits.len() == 1 {
+                write!(fmt, "Skipping rebase of commit ")?;
+                tx.write_commit_summary(fmt.as_mut(), &skipped_commits[0])?;
+                writeln!(fmt)?;
+            } else {
+                writeln!(fmt, "Skipping rebase of commits:")?;
+                for commit in skipped_commits {
+                    write!(fmt, "  ")?;
+                    tx.write_commit_summary(fmt.as_mut(), &commit)?;
+                    writeln!(fmt)?;
+                }
+            }
```

Comment:
I realized that since one usage is outside of a transaction whilst the other is within, I can't reuse the same function for both. Not sure if there are any workarounds that I'm not aware of.

Reply:
Since these commits are skipped, I thought we wouldn't have to use the post-rebased repo state. If that matters, we can't use the base repo/workspace_helper.

Reply:
The issue I ran into is that `log_skipped_rebase_commits_message` accepts `&WorkspaceCommandHelper`, but the `workspace_command` can't be borrowed immutably since it's already borrowed mutably for the lifetime of the transaction. I guess the message printing can be shifted outside the transaction also?

Reply:
Maybe this is where `tx.base_workspace_helper()` is useful?

Reply:
Thanks for the pointer, didn't know about that function 😅  Updated!